docs/html/reference/pip_install.rst

```diff
@@ -63,7 +63,50 @@ explicitly state the project name.
 Satisfying Requirements
 -----------------------
 
-Once pip has the set of requirements to satisfy, it chooses which version of
+In order to satisfy the requirements specified on the command line, pip
+calculates a set of packages and versions that meets the given requirements. The
+details of this process are as follows:
+
+1. Pip considers all packages named in requirements specified on the command
+   line, and their dependencies, recursively.
+2. No other packages are considered when determining what to install. In
+   particular, installed packages not part of the dependency tree constructed
+   in step 1 will not be considered.
+3. When looking for solutions to the various version constraints, pip will
+   normally prioritise the installed version of a given package over any other.
+   However, note that this does not mean that pip will certainly retain that
+   version - other constraints may influence this.
+4. The ``--upgrade`` flag alters this behaviour by *not* prioritising installed
+   versions for packages named on the command line. This is the default,
+   "only-if-needed" upgrade strategy. It is also possible to specify
+   ``--upgrade-strategy=eager``, which doesn't prioritise installed versions
+   for *any* installed package - directly specified or dependency.
+5. Apart from the above exceptions around installed versions, pip will
+   prioritise later versions over earlier versions.
+
+Once pip has satisfied all constraints, it will install the selected versions,
+uninstalling existing installed versions if necessary.
+
+It should be noted that in order to satisfy the given requirements, pip will
+upgrade (or even downgrade) installed packages as needed, regardless of whether
+the ``--upgrade`` flag is present. The ``--upgrade`` flag only influences the
+*choice* of what version to install, not whether to allow upgrades.
```

Comment:
This sentence feels so confusing. I can instinctively understand the implication, but feel it is not explaining it right, *but* can’t point my finger to where nor suggestion something better 😞

Reply:
Agreed, it's pretty bad. I need to rethink this (at least part of the problem is the fact that `--upgrade` doesn't do what I (and I suspect most people) think it should.

I'll have another go at the text here.

Reply:
TBH as an user after reading this I have no idea what `--upgrade` is supposed to do anymore.  With the legacy resolver, upgrading everything as a whole was not possible and I *think* the flag was to enable users to resolve dependencies by themselves and decide which package to upgrade (i.e. `pip install -U this` is equivalent to `upgrade this` in other package managers).  With the new resolver, it might be less mind-boggling for users if we have `pip upgrade`, and just let the resolver tries its best during installation.

Since `--upgrade` doesn't really preserve backward-compatibility (it would if `--upgrade-strategy=eager` is provided IIUC), IMHO if when the new resolver becomes default and `upgrade` subcommand is yet to be ready, we may want to flag it as deprecation (e.g. *`--upgrade` may not do what you want, since the dep-resolve process influences the choice of which package to be upgraded/downgraded).  If one wants some sort of a new version, perse can always specify it explicitly, which also complies with the Zen of Python.

Reply:
> `pip install -U this` is equivalent to `upgrade this` in other package managers

I think this is the correct mental model both with the legacy and new resolvers. But the problem is in the details. What does `upgrade this` actually mean? Does it only upgeade `this`? Does it upgrade `this` and all its dependencies? How does “upgrade all its dependencies” even mean?

The operation “upgrade” actually operates in a different abstraction level from package installation, and most “package managers” implement it as such: in a different abstraction level with the manifest/lock mechanism. Having an upgrade operation in the abstraction level pip operates in is quirky by itself, so the definition has to be quirky as a result.

Reply:
> TBH as an user after reading this I have no idea what --upgrade is supposed to do anymore. 

TBH after reading the code I lost all sense of what `--upgrade` does, so I'm glad that I captured my understanding so accurately 🙂

More seriously, yes, once you start digging pip's upgrade operation is a very weird beast and it's not at all what youthink it is at first glance. One thing that became obvious to me, for example, while writing tests for the new upgrade strategy code, is that upgrade doesn't even have any meaning unless you factor in the concept of "a new version has been released" and that is an idea that's never really relevant anywhere else in pip.

My feeling is that there's a fairly significant restructuring that needs to be done here before it'll be possible to describe things really accurately. What I'm not sure about is whether we'll be able to get a better description *before* we do that, or whether we'd be better ignoring the user documentation problem for now and moving this description into the internals documentation (where it can afford to stay a bit confusing...).

But thanks for all the thoughtful responses, I'll let the discussion continue for a while and then read it all up at once, rather than keep responding bit-by-bit like this.